src/backend/distributed/planner/multi_explain.c

```diff
@@ -639,6 +688,289 @@ BuildRemoteExplainQuery(char *queryString, ExplainState *es)
+{
+	/*
+	 * Make sure stats accumulation is done.  (Note: it's okay if several
+	 * levels of hook all do this.)
+	 */
+	InstrEndLoop(queryDesc->totaltime);
+
+	ExplainState *es = NewExplainState();
+
+	es->analyze = true;
+	es->verbose = WorkerQueryExplainOptions.verbose;
+	es->buffers = WorkerQueryExplainOptions.buffers;
+	es->timing = WorkerQueryExplainOptions.timing;
+	es->summary = WorkerQueryExplainOptions.summary;
+	es->format = WorkerQueryExplainOptions.format;
+	es->costs = WorkerQueryExplainOptions.costs;
+
+	ExplainBeginOutput(es);
+
+	if (WorkerQueryExplainOptions.query)
+	{
+		ExplainQueryText(es, queryDesc);
+	}
+
+	ExplainPrintPlan(es, queryDesc);
+
+	if (es->costs)
+	{
+		ExplainPrintJITSummary(es, queryDesc);
+	}
+
+	ExplainEndOutput(es);
+
+	/*
+	 * We don't always send BEGIN/END to workers for all queries. Changing
+	 * transaction management for EXPLAIN ANALYZE can change their runtime
+	 * metrics, so we don't want to change it for more realistic EXPLAIN
+	 * ANALYZE.
+	 *
+	 * Therefore, SavedExplainPlan might be fetched in a different transaction
+	 * than it was created at. So we allocate it in TopMemoryContext so it
+	 * survives transaction boundaries.
+	 */
+	MemoryContext oldContext = MemoryContextSwitchTo(TopMemoryContext);
+
+	if (SavedExplainPlan != NULL)
+	{
+		pfree(SavedExplainPlan);
+	}
+
+	SavedExplainPlan = pstrdup(es->str->data);
+
+	MemoryContextSwitchTo(oldContext);
+
+	pfree(es->str->data);
+}
+
+
+/*
+ * RequestedForExplainAnalyze returns true if we should get the EXPLAIN ANALYZE
+ * output for the given custom scan node.
+ */
+bool
+RequestedForExplainAnalyze(CustomScanState *node)
+{
+	return (node->ss.ps.state->es_instrument != 0);
+}
+
+
+/*
+ * InstallExplainAnalyzeHooks installs hooks on given tasks so EXPLAIN ANALYZE
+ * output of them are saved & fetched from workers.
+ */
+void
+InstallExplainAnalyzeHooks(List *taskList)
+{
+	Task *task = NULL;
+	foreach_ptr(task, taskList)
+	{
+		task->preExecutionHook = ExplainAnalyzePreExecutionHook;
+		task->postExecutionHook = ExplainAnalyzePostExecutionHook;
+	}
+}
+
+
+/*
+ * ExplainAnalyzePreExecutionHook implements task->preExecutionHook and sends
+ * EXPLAIN ANALYZE params over the given connection.
+ */
+static void
+ExplainAnalyzePreExecutionHook(Task *task, int placementIndex,
+							   MultiConnection *connection)
+{
+	StringInfo query = makeStringInfo();
+	appendStringInfo(query,
+					 "SELECT worker_save_query_explain_analyze(true, %s, %s, %s, %s, %s, %d)",
+					 CurrentDistributedQueryExplainOptions.verbose ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.costs ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.timing ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.summary ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.query ? "true" : "false",
+					 (int) CurrentDistributedQueryExplainOptions.format);
+
+	ExecuteCriticalRemoteCommand(connection, query->data);
```

Comment:
Doing this synchronously adds quite a bit of latency to the whole execution. I'm wondering whether we should correct for that, or avoid it by batching the command together with the actual SQL command.

Reply:
I'm not sure how useful this is:
```
select count(*) from test;
┌───────┐
│ count │
├───────┤
│   100 │
└───────┘
(1 row)

Time: 52.762 ms

explain analyze select count(*) from test;
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                              QUERY PLAN                                                              │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Aggregate  (cost=250.00..250.02 rows=1 width=8) (actual time=307.353..307.354 rows=1 loops=1)                                        │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8) (actual time=307.296..307.328 rows=80 loops=1)             │
│         Task Count: 80                                                                                                               │
│         Tasks Shown: One of 80                                                                                                       │
│         ->  Task                                                                                                                     │
│               Node: host=localhost port=9700 dbname=postgres                                                                         │
│               ->  Aggregate  (cost=38.25..38.26 rows=1 width=8) (actual time=0.017..0.017 rows=1 loops=1)                            │
│                     ->  Seq Scan on test_102008 test  (cost=0.00..32.60 rows=2260 width=0) (actual time=0.011..0.012 rows=1 loops=1) │
│ Planning Time: 1.597 ms                                                                                                              │
│ Execution Time: 309.319 ms                                                                                                           │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(10 rows)

Time: 315.501 ms
```

Reply:
For comparison, on master:
```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                              QUERY PLAN                                                              │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Aggregate  (cost=250.00..250.02 rows=1 width=8) (actual time=51.516..51.516 rows=1 loops=1)                                          │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8) (actual time=51.086..51.272 rows=80 loops=1)               │
│         Task Count: 80                                                                                                               │
│         Tasks Shown: One of 80                                                                                                       │
│         ->  Task                                                                                                                     │
│               Node: host=localhost port=9700 dbname=postgres                                                                         │
│               ->  Aggregate  (cost=38.25..38.26 rows=1 width=8) (actual time=0.041..0.041 rows=1 loops=1)                            │
│                     ->  Seq Scan on test_102008 test  (cost=0.00..32.60 rows=2260 width=0) (actual time=0.027..0.029 rows=1 loops=1) │
│                   Planning Time: 0.145 ms                                                                                            │
│                   Execution Time: 0.184 ms                                                                                           │
│ Planning Time: 1.895 ms                                                                                                              │
│ Execution Time: 54.274 ms                                                                                                            │
└─────────────────────────
````

Reply:
Today I worked on this, and have a WIP solution, which still needs some work: https://github.com/citusdata/citus/commit/98cf2dbb6a642281151c95d6af35165caf1633d2

The idea is if asked for EXPLAIN ANALYZE, add the set/fetch UDF calls as prefix and suffix of the task query. Then adaptive executor calls `ExplainAnalyzeResultConsumerHook` when receiving results, and if a result belongs to EXPLAIN ANALYZE udf calls, multi_explain processes it. So not much synchronicity anymore. 

It worked fine & fast for few cases I tried locally, but fails regression tests, so I will continue working on it tomorrow.

Agreed with rest of your comments. I'll address them once this is done.

Reply:
This hook is quite interesting. It does feel a little clunky with the magic strings. Another case where we did something similarly weird is insert..select with re-partitioning where we include the node ID in the query just to be able to determine which node the result came from later.

One approach could be to generalize away from the tuple store, e.g. instead of:

```c
tuplestore_puttuple(tupleStore, heapTuple);
```

we do:
```c
executorTupleDestination->putTuple(executorTupleDestination, task, placementIndex, queryNumber, heapTuple);
```

The normal executor tuple destination would simply call `tuplestore_puttuple`. In case of EXPLAIN ANALYZE, we wrap it in a function that ignores queryNumber 0, calls the default putTuple for queryNumer 1, and stores the explain output for queryNumber 2.

Alternatively, we only do this when the hook is set.

Not sure whether this makes more sense as an executor-level or task/job-level hook. It might be the latter if we want to start thinking of the executor as a command-running device rather than a tuple-collecting device.

Reply:
Something to watch out for:
```sql
PREPARE test(int) AS with a as (update test set y = 0 where x = $1 returning *) select * from a;
EXPLAIN ANALYZE EXECUTE test(3);
ERROR:  cannot insert multiple commands into a prepared statement
CONTEXT:  while executing command on localhost:5432
```

Reply:
Foundation of the tuple destination approach is implemented in https://github.com/citusdata/citus/pull/3871. I'll try to send that PR for review tomorrow. I sent a meeting invite to discuss decisions there.